### PR TITLE
feat(server): item infrastructure

### PR DESCRIPTION
**Pull request by mimoham24**
# Overview

### CODE REVIEW

File: server/internal/infrastructure/memory/item.go [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+	"github.com/reearth/reearth-cms/server/pkg/id"
+	"github.com/reearth/reearth-cms/server/pkg/item"
+	"github.com/reearth/reearthx/rerror"
+	"github.com/reearth/reearthx/util"
+	"golang.org/x/exp/slices"
+)
+
+type Item struct {
+	data *util.SyncMap[item.ID, *item.Item]
+	err  error
```
Comment (rot1024) [inline server/internal/infrastructure/memory/item.go:16]:
```suggestion
	data *memorygit.VersionedSyncMap[item.ID, *item.Item]
```

use `memorygit.VersionedSyncMap`

File: server/pkg/item/item.go [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+	schemaID schema.ID
+	fields   []*Field
+}
+
+func (i *Item) ID() ID {
+	return i.id
+}
+
+func (i *Item) Fields() []*Field {
+	return i.fields
```
Comment (rot1024) [inline server/pkg/item/item.go:21]:
```suggestion
	return slices.Clone(i.fields)
```

File: server/pkg/item/list.go [hunk 3]
```diff
@@ -0,0 +1,15 @@
+package item
+
+import (
+	"golang.org/x/exp/slices"
+)
+
+type List []*Item
```
Comment (rot1024) [inline server/pkg/item/list.go:7]: Is it used?

File: server/pkg/item/field.go [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+	valueType     schema.Type
+	value         any
+}
+
+func NewField(schemaFieldID schema.FieldID, valueType schema.Type, value any) *Field {
+	return &Field{schemaFieldID: schemaFieldID, valueType: valueType, value: value}
+}
+
+func (f *Field) SchemaFieldID() schema.FieldID {
+	return f.schemaFieldID
```
Comment (rot1024) [inline server/pkg/item/field.go:16]: add test

File: server/pkg/item/field.go [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+}
+
+func (f *Field) ValueType() schema.Type {
+	return f.valueType
```
Comment (rot1024) [inline server/pkg/item/field.go:20]: add test

File: server/pkg/item/field.go [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+}
+
+func (f *Field) Value() any {
+	return f.value
```
Comment (rot1024) [inline server/pkg/item/field.go:24]: add test

File: server/pkg/item/builder.go [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+	return b
+}
+
+func (b *Builder) NewID() *Builder {
+	b.i.id = NewID()
+	return b
+}
+
+func (b *Builder) Fields(fs []*Field) *Builder {
+	b.i.fields = fs
```
Comment (rot1024) [inline server/pkg/item/builder.go:31]:
```suggestion
	b.i.fields = slices.Clone(fs)
```

File: server/internal/infrastructure/mongo/mongodoc/item.go [hunk 8]
```diff
@@ -0,0 +1,102 @@
...
+	if err := bson.Unmarshal(raw, &doc); err != nil {
+		return err
+	}
+	item, err := doc.Model()
+	if err != nil {
+		return err
+	}
+	c.Rows = append(c.Rows, item)
+	return nil
+}
```
Comment (rot1024) [inline server/internal/infrastructure/mongo/mongodoc/item.go:40]: See https://github.com/reearth/reearth-cms/blob/main/server/internal/infrastructure/mongo/mongodoc/model.go#L63

```suggestion
type ItemConsumer = mongox.SliceFuncConsumer[*ItemDocument, *item.Item]

func NewItemConsumer() *ItemConsumer {
	return NewComsumer[*ItemDocument, *item.Item]()
}
```

**Review by rot1024 — CHANGES_REQUESTED** [review]

File: server/internal/infrastructure/mongo/item.go [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+type itemRepo struct {
+	client *mongogit.Collection
+}
+
+func NewItem(client *mongox.Client) repo.Item {
+	return &itemRepo{client: mongogit.NewCollection(client.WithCollection("item"))}
+}
+
+func (i itemRepo) FindByIDs(ctx context.Context, ids id.ItemIDList) ([]*item.Item, error) {
+	c := &mongox.SliceConsumer[mongodoc.ItemDocument]{}
```
Comment (rot1024) [inline server/internal/infrastructure/mongo/item.go:25]:
```suggestion
	c := mongox.NewItemConsumer()
```

File: server/internal/infrastructure/mongo/item.go [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+		if err != nil {
+			return nil, err
+		}
+		res = append(res, it)
+	}
+	return res, nil
+}
+
+func (i itemRepo) FindByID(ctx context.Context, id id.ItemID) (*item.Item, error) {
+	c := &mongox.SliceConsumer[mongodoc.ItemDocument]{}
```
Comment (rot1024) [inline server/internal/infrastructure/mongo/item.go:45]:
```suggestion
	c := mongox.NewItemConsumer()
```

**Review by rot1024 — CHANGES_REQUESTED** [review]

File: server/pkg/item/list.go (hunk 3 above)
Comment (mimoham24) [reply server/pkg/item/list.go:7]: yes, it's used in mongo layer

File: server/internal/infrastructure/mongo/item_test.go [hunk 11]
```diff
@@ -0,0 +1,83 @@
...
+	init := mongotest.Connect(t)
+	client := mongox.NewClientWithDatabase(init(t))
+	repo := NewItem(client)
+	ctx := context.Background()
+	_ = repo.Save(ctx, i1)
+	err := repo.Remove(ctx, i1.ID())
+	assert.NoError(t, err)
+	got, err := repo.FindByID(ctx, i1.ID())
+	assert.Nil(t, got)
+	assert.Error(t, err)
```
Comment (rot1024) [inline server/internal/infrastructure/mongo/item_test.go:82]:
```suggestion
	assert.Equal(t, rerror.ErrNotFound, err)
```

File: server/internal/infrastructure/mongo/mongodoc/item.go [hunk 12]
```diff
@@ -0,0 +1,103 @@
...
+	"github.com/reearth/reearth-cms/server/pkg/version"
+	"github.com/reearth/reearthx/mongox"
+)
+
+type ItemDocument struct {
+	ID     string
+	Schema string
+	Fields []ItemFieldDoc
+}
+type ItemFieldDoc struct {
```
Comment (rot1024) [inline server/internal/infrastructure/mongo/mongodoc/item.go:17]:
```suggestion

type ItemFieldDoc struct {
```

File: server/internal/infrastructure/mongo/item_test.go [hunk 13]
```diff
@@ -0,0 +1,83 @@
...
+}
```
Comment (rot1024) [inline server/internal/infrastructure/mongo/item_test.go:83]: `FindBySchema`, `FindAllVersionsByID`, `FindByIDs` are not tested. Can you add unit tests for them?

**Review by rot1024 — CHANGES_REQUESTED** [review]

**Review by rot1024 — APPROVED** [review]

**Pull request merged**